sync/tc.py

```diff
@@ -220,32 +225,63 @@ def download_logs(self, destination, file_names, retry=5):
         if not file_names:
             return []
 
-        urls = [ARTIFACTS_BASE + "{task}/{run}/public/test_info//%s" % file_name
-                for file_name in file_names]
         logger.info("Downloading logs to %s" % destination)
-        for task in self.tasks:
-            status = task.get("status", {})
-            for run in status.get("runs", []):
-                if "_log_paths" not in run:
-                    run["_log_paths"] = {}
-                for url in urls:
-                    params = {
-                        "task": status["taskId"],
-                        "run": run["runId"],
-                    }
-                    params["file_name"] = url.rsplit("/", 1)[1]
-                    log_url = url.format(**params)
-                    log_name = "{task}_{run}_{file_name}".format(**params)
-                    success = False
-                    logger.debug("Trying to download {}".format(log_url))
-                    log_path = os.path.abspath(os.path.join(destination, log_name))
-                    if not os.path.exists(log_path):
-                        success = download(log_url, log_path, retry)
-                    else:
-                        success = True
-                    if not success:
-                        logger.warning("Failed to download log from {}".format(log_url))
-                    run["_log_paths"][params["file_name"]] = log_path
+        t0 = time.time()
+        executor = ThreadExecutor(8, work_fn=get_task_artifacts, init_fn=start_session)
+        errors = executor.run([((), {
+            "destination": destination,
+            "task": item,
+            "file_names": file_names,
+            "retry": retry
+        }) for item in self.tasks])
+        # TODO: not sure if we can avoid tolerating some errors here, but
+        # there is probably some sign of badness less than all the downloads
+        # erroring
+        for error in errors:
+            logger.warning(traceback.format_exc(error))
+        if len(errors) == len(self.tasks):
+            raise RetryableError("Downloading logs all failed")
+        logger.info("Downloading logs took %s" % (time.time() - t0))
+
+
+def start_session():
+    return {"session": requests.Session()}
+
+
+def get_task_artifacts(destination, task, file_names, session, retry):
+    status = task.get("status", {})
+    if not status.get("runs"):
+        logger.debug("No runs for task %s" % status["taskId"])
+        return
+    artifacts_base_url = QUEUE_BASE + "task/%s/artifacts" % status["taskId"]
+    try:
+        artifacts = fetch_json(artifacts_base_url, session=session)
+    except requests.HTTPError as e:
+        logger.warning(e.message)
+    artifact_urls = ["%s/%s" % (artifacts_base_url, item["name"])
+                     for item in artifacts["artifacts"]
+                     if any(item["name"].endswith("/" + file_name)
+                            for file_name in file_names)]
+
+    run = status["runs"][-1]
```

Comment:
I'm assuming this works, but it's notable to me that now we're just looking at one run instead of a list of runs. I guess there's always just one run per task and a retriggered job is in an independent task? I don't remember the structure of the task status. Pointing it out in case there was an oversight.

Reply:
Yeah, so a retrigger is not a rerun, it's a different task. reruns are mostly the automatic restarts in cases where the spot instance is discarded or whatever. I think the assumption that the final run is always the useful one is pretty reasonable (the fetchlogs tool assumes this and I haven't noticed it causing a problem yet, so there's also some empirical evidence that the theory works). The advantage of not looking at all the runs is obviously that it's a bit faster.